File: Utils/Cpp_Samples/lbm2bmp.cpp
///////////////////////////////////////////////////////////////////////////////
// FuryUtils Sample
//
// Converting an Lbm format file to a windows Bmp format file
//
// using:
//
//    Lbm
//       class for reading / writing images in Lbm format.
//    Bmp
//       class for reading / writing images in Bmp format.
//    Exceptions::Exception
//       exception class for reporting errors.
//
//
//   Suggested inputs from the testassets: pal8out.lbm
//
///////////////////////////////////////////////////////////////////////////////

#include <ios>
#include <fstream>
#include "../include/FuryUtils.hpp"

void lbm2bmp(const char * lbmFileName, const char * bmpFileName) {
	printf("lbm2bmp Sample\n");
	
	try
	{
		// Get a stream for the Lbm format file and load it into a buffer
		std::ifstream lbmStream(lbmFileName, std::ios::binary | std::ios::ate);
		std::streamsize size = lbmStream.tellg();
		lbmStream.seekg(0, std::ios::beg);

		std::vector<uint8_t> lbmBuffer((uint32_t)size);
		lbmStream.read((char *)(lbmBuffer.data()), size);

		// Create an Lbm object from the input stream
		FuryUtils::Image::Lbm lbm(lbmBuffer);
		
		// Create a Bmp object from the Lbm
		FuryUtils::Image::Bmp bmp(lbm);

		// Get a buffer of the image in Bmp format and save it to a stream
		std::vector<uint8_t> outBuffer;
		bmp.Buffer(outBuffer);

		std::ofstream outFile(bmpFileName, std::ios::binary | std::ios::trunc);
		outFile.write((char *)(outBuffer.data()), outBuffer.size());

		return;
	}
	catch (FuryUtils::Exceptions::Exception e)
	{
		printf("Error:\n\n%d %s\n", e._errorCode, e._errorString.c_str());
		return;
	}
	return;
}

File: CS_Samples/lbm2bmp.cs
///////////////////////////////////////////////////////////////////////////////
///
/// FuryUtils Sample
/// 
/// Converting an Lbm format file to a windows Bmp format file.
/// 
/// using:
/// 
///    Bmp
///         class for reading/writing images in Bmp format.
///    Lbm
///         class for reading/writing images in Lbm format.
///    FuryException
///         exception class for reporting errors.
///
///
/// Suggested inputs from the testassets: pal8out.lbm
/// 
///////////////////////////////////////////////////////////////////////////////

using carbon14.FuryStudio.Utils;

namespace carbon14.FuryStudio.CS_Samples
{
    internal class lbm2bmp
    {
        public static void Execute(string lbmFileName, string bmpFileName)
        {
            Console.WriteLine("lbm2bmp Sample");

            try
            {
                // Read lbm file into buffer.
                byte[] lbmBuffer;
                using (FileStream fs = new FileStream(lbmFileName, FileMode.Open))
                {
                    lbmBuffer = new byte[fs.Length];
                    fs.Read(lbmBuffer, 0, lbmBuffer.Length);
                }

                byte[]? bmpBuffer;

                // Create an Lbm object from the buffer.
                // Create a Bmp object from that
                // and get buffers for the raw image and palette data.
                using (Lbm lbm = new Lbm(lbmBuffer))
                {
                    using (Bmp bmp = new Bmp(lbm))
                    {
                        bmpBuffer = bmp.Buffer;
                    }
                }

                // Write the bmp data into a file.
                if (bmpBuffer != null)
                {
                    using (FileStream fs = new FileStream(bmpFileName, FileMode.Create))
                    {
                        fs.Write(bmpBuffer, 0, bmpBuffer.Length);
                    }
                }
            }
            catch (FuryException ex)
            {
                Console.WriteLine($"Error: {ex.ErrorCode} {ex.Message}");
            }
        }
    }
}
